Update cat base to handle redirects properly

chuckreynolds opened the pull request:
Updated to handle redirects properly. 
It was doing /blog/category/foocat/ => /foocat and 404'ing...  This patch should handle all/most? (test) redirects accordingly; example /blog/category/foocat/ => /blog/foocat/

chuckreynolds committed:
Update cat base to handle redirects properly

Updated to handle redirects properly. 
It was doing /blog/category/foocat/ => /foocat and 404'ing... 
This patch should handle all/most? (test) redirects accordingly; example /blog/category/foocat/ => /blog/foocat/

sybrew commented on permalinks/tsf-remove-category-base.php:
```diff
@@ -101,10 +101,16 @@ function redirect_base( $query_vars ) {
-	);
+	// Determine blog prefix, similar to modify_category_rewrite_rules
+	$permalink_structure = \get_option( 'permalink_structure' );
+	$blog_prefix = \is_main_site() && str_starts_with( $permalink_structure, '/blog/' )
```
Thanks for the PR!

Question: Why are we testing `is_main_site()` here?

Perhaps we should use `home_url( '/' )` and let WordPress figure out the base, instead of using `get_option( 'home' )` (at line 111).

chuckreynolds replied:
why is_main_site()... idk. kinda vibe coded it while drinking on vacation :) (_only time i have to work on non-work stuff_) Probably don't need that. GPT4.1 probably saw https://github.com/sybrew/tsf-snippets/blob/9803737383a9b3276c2bf6cc0eccd6d7d45869a4/permalinks/tsf-remove-category-base.php#L127 in modify_category_rewrite_rules() and duped that functionality. 
It also feels a little 'hard coded' to the use case of `.tld/blog/**`

chuckreynolds committed:
Update redirect_base() in tsf-remove-category-base.php

Use get_category_link() to get the correct URL for the category by slug.
Use the slug from the redirect match to get the term object.
Redirect to the canonical category URL using get_category_link()

chuckreynolds commented:
New way to handle it. Getting cat url/slug instead as it already is handled correctly so may as well use that instead of recalcing what the slug should be. @sybrew 

Edit 1: That said I still think this in `modify_category_rewrite_rules()` is hard-fixed to /blog/ being in the permalink structure. 
```
$blog_prefix = \is_main_site() && str_starts_with( $permalink_structure, '/blog/' )
    ? 'blog/'
    : '';
```
That'd need to be tackled if this were to be released more publicly. 

Edit 2: Looking at YxxSEO plugin that's what they have; so is_multisite came from that too in /inc/class-rewrite.php
```
$permalink_structure = get_option( 'permalink_structure' );

$blog_prefix = '';
if ( strpos( $permalink_structure, '/blog/' ) === 0 ) {
	if ( ( is_multisite() && ! is_subdomain_install() ) || is_main_site() || is_main_network() ) {
		$blog_prefix = 'blog/';
	}
}
```

sybrew commented on permalinks/tsf-remove-category-base.php:
```diff
@@ -91,23 +91,37 @@ function register_query_vars( $query_vars ) {
+	}
+
+	// Fallback: redirect to home if not found
+	wp_safe_redirect( home_url(), 301 );
```
Never do this! 404s are good. I smell a bit too much vibe 😅

sybrew commented on permalinks/tsf-remove-category-base.php:
```diff
@@ -91,23 +91,37 @@ function register_query_vars( $query_vars ) {
+/**
+ * Helper: Get category by slug.
+ */
+function get_category_by_slug( $slug ) {
```
This function already exists in Core: https://developer.wordpress.org/reference/functions/get_category_by_slug/. It's heavy, though. If possible, I'd rather trim out the slug outright instead of performing category lookups.
In other words, we probably just want to modify the rewrite base so that `/%category%/` disappears.

chuckreynolds replied:
Have at it... you're a better coder than me most likely... i'm out of the game mostly lol. Too much dust for me to knock off to code full-time anymore. I just want a solution that's solid and can be shared back to the community for everybody to use.

sybrew commented:
Well, that was a waste of $390 🥲, I thought Copilot would consider our conversation and amend the commit accordingly.

sybrew commented:
I'll look at this again later (I'm extremely swamped) -- if I were confident that this feature would be foolproof, it'd have been implemented in TSF a long time ago. 

Due to performance issues I mentioned in the review, I'm skeptical about the PR as it stands, even though the suggestion you're making could make it much more robust.

I'll keep this PR open for now, because it does solve a problem, though I need to get my head into it to figure out if it's _the best_ and _the only_ solution. Tweaking is necessary, such as removing the fallback redirect, but that can always happen later.
